src/views/Referrals/Referrals.js

```diff
@@ -0,0 +1,601 @@
+import React, { useEffect, useState } from "react";
+import { useWeb3React } from "@web3-react/core";
+
+import Card from "../../components/Common/Card";
+import SEO from "../../components/Common/SEO";
+import Tab from "../../components/Tab/Tab";
+import Footer from "../../Footer";
+import {
+  useChainId,
+  getPageTitle,
+  formatAmount,
+  USD_DECIMALS,
+  helperToast,
+  formatDate,
+  getTokenInfo,
+  getExplorerUrl,
+  shortenAddress,
+  ARBITRUM,
+  bigNumberify,
+  REFERRAL_CODE_QUERY_PARAMS,
+  isHashZero,
+} from "../../Helpers";
+import { decodeReferralCode, useReferralsData } from "../../Api/referrals";
+
+import "./Referrals.css";
+import { registerReferralCode, setTraderReferralCodeByUser, useInfoTokens, useUserReferralCode } from "../../Api";
+import { utils } from "ethers";
+import { BiCopy, BiEditAlt } from "react-icons/bi";
+import Tooltip from "../../components/Tooltip/Tooltip";
+import { useCopyToClipboard } from "react-use";
+import Loader from "../../components/Common/Loader";
+import Modal from "../../components/Modal/Modal";
+import { RiQuestionLine } from "react-icons/ri";
+import { FiPlus } from "react-icons/fi";
+
+const getSampleReferrarStat = (code) => {
+  return {
+    discountUsd: bigNumberify(0),
+    referralCode: code,
+    totalRebateUsd: bigNumberify(0),
+    tradedReferralsCount: 0,
+    trades: 0,
+    volume: bigNumberify(0),
+  };
+};
+
+const TRADERS = "Traders";
+const AFFILIATES = "Affiliates";
+let TAB_OPTIONS = [TRADERS, AFFILIATES];
+
+function getDollarValue(value) {
+  return `$${formatAmount(value, USD_DECIMALS, 2, true, "0.00")}`;
+}
+
+function getErrorMessage(value) {
+  let trimmedValue = value.trim();
+  let invalid = /\s/;
+  if (invalid.test(trimmedValue)) {
+    return "The referral code can't contain spaces.";
+  }
+  return "";
+}
+
+function Referrals({ connectWallet, setPendingTxns, pendingTxns }) {
+  const { active, account, library } = useWeb3React();
+  const { chainId } = useChainId();
+  const { infoTokens } = useInfoTokens(library, chainId, active, undefined, undefined);
+  let [activeTab, setActiveTab] = useState(TRADERS);
+  const referralsData = useReferralsData(chainId, account);
+  const [referralsDataState, setReferralsDataState] = useState(referralsData);
+  const { userReferralCode } = useUserReferralCode(library, chainId, account);
+
+  useEffect(() => {
+    setReferralsDataState(referralsData);
```

Comment:
mmm thinking if it can work unexpectedly in some case.

so here you're setting "initial" value when data from subgraph is loaded
then when user adds code `referralsDataState` is updated
in theory if `referralsData` changes it can override that updated state 🤔

Reply:
> I was wondering if apolloClient gives a way to re-fetch the data programmatically.

we may probably use `useSWR` there to make it more flexible. or as a simple solution we may pass some `key` to use useUserReferralCode to use in `useEffect`'s dependencies

atm all subgraph data in UI is just loaded once and never updates. probably it's not critical, i don't think it makes sense  to update rebates data realtime (it doesn't change that often) 🤔

Reply:
Yeah, because of that reason only I though to just use optimistic update. Because when a user adds new referral code it's not going to have any volume. So in future when they refresh the page they will get the updated data.